### PR TITLE
csv-source-streams

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

File: src/source/csv.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+use futures::stream::{self, BoxStream, StreamExt};
+use geoengine_datatypes::collections::PointCollection;
+use geoengine_datatypes::primitives::{Coordinate, TimeInterval};
+use serde::{Deserialize, Serialize};
+use snafu::{ensure, OptionExt, ResultExt};
+use std::fs::File;
+use std::i64;
+use std::path::PathBuf;
+
+/// Parameters for the GDAL Source Operator
```
Comment (michaelmattig) [inline src/source/csv.rs]:
```suggestion
/// Parameters for the CSV Source Operator
```

**Comment by michaelmattig** [comment]
looks quite nice. Two questions:
1. is there any advantage using streams over iterators here? Is there anything being asynchronously processed?
2. does it make sense to chunk the data set by number of rows? Rows can be very small or very big...

**Comment by ChristianBeilschmidt** [comment]
> looks quite nice. Two questions:
> 
>     1. is there any advantage using streams over iterators here? Is there anything being asynchronously processed?
The Iterator that reads the file is now independent of the usage of the PointCollection chunk. This means one could call `.buffered(…)` so that the next chunk is fetched (from the file system) while the PointCollection is processed in a (potentially) different thread.

> 
>     2. does it make sense to chunk the data set by the number of rows? Rows can be very small or very big...
This is a good point. So there is a trade-off. On the one hand, using something like 1MB per chunk evens the exact workload. On the other hand, the chunks would vary in size.

We have to answer the questions whether we want to ensure that the chunks are, for instance, (100, 100, 30) instead of (20, 140, 70) -- this means the chunks are filled up until the last one.

Furthermore, I guess it is much more complicated to use the actual byte size of chunks.

We have to think about this.

**Comment by michaelmattig** [comment]
>The Iterator that reads the file is now independent of the usage of the PointCollection chunk. This means one could call .buffered(…) so that the next chunk is fetched (from the file system) while the PointCollection is processed in a (potentially) different thread.
true, I didn't realize the csv reader reads the file lazily

> This is a good point. So there is a trade-off. On the one hand, using something like 1MB per chunk evens the exact workload. On the other hand, the chunks would vary in size.
We have to answer the questions whether we want to ensure that the chunks are, for instance, (100, 100, 30) instead of (20, 140, 70) -- this means the chunks are filled up until the last one.
Furthermore, I guess it is much more complicated to use the actual byte size of chunks.
We have to think about this.

it's not even clear, that two chunks of the same size incur the same workload if the complexity of the operation depends on the actual data. Once, we have a running system we can investigate this further ;)

From my side you can continue the implementation of the operator.

**Pull request merged**